### PR TITLE
benchmark: support encoded output for benchmark result

### LINKED ISSUE #11459: benchmark: add file output option to benchmark tool
Currently the benchmark tool outputs performance summary and details to console in human readable format, but the results cannot be parsed programmatically.

The proposal here is to add an option to output to a file specified by user input, potentially in JSON format.

This output file option opens the opportunity for automated analysis or data visualization.

This will also be prerequisite for #6002

### PR DESCRIPTION
**Pull request by anson627**
this PR adds support for encoded output for benchmark results. Only json format is supported currently. If output format is set and valid, only raw objects of benchmark results are written to stdout, anything else written to stdout (e.g. progress bar) is suppressed in order to prevent breaking output parsing. This provides more flexibility: users can redirect the stdout to a file if they wish, instead of forcing the output to a file.

Fixes #11459

### CODE REVIEW

**Comment by xiang90** [comment]
can you just add an output format instead of forcing the output to a file? user can redirect the io to a file if they wish.

**Comment by anson627** [comment]
I actually tried this approach first, it doesn't work well, because the existing implementation doesn't have a unified way to control what can be written to the output.

For example, the progress bar for user to monitor the progress of the benchmark tests is also written to the console output. If we mute the progress bar for json output, it makes it hard for users to determine if benchmark is hanging or making progress. Even if we decide to mute it, implementation wise we will have to check and mute progress bar at multiple places:
https://github.com/etcd-io/etcd/blob/master/tools/benchmark/cmd/put.go#L89
and 
https://github.com/etcd-io/etcd/blob/master/tools/benchmark/cmd/put.go#L132
for each subcommand. 

It also means it is fragile, if there is a printing line anywhere added in future without checking, the json output parsing can break. If there are low level errors (for example, grpc error) printed out to console, the json output parsing is broken.

**Comment by anson627** [comment]
ping

**Comment by anson627** [comment]
ping @gyuho

**Comment by xiang90** [comment]
if json format is set, we can simply not output the progress bar to stdout. Basically anything that is not json will not go to stdout.

**Comment by anson627** [comment]
just want to make sure I understand correctly.

If json format is set, at begin of the command execution, we can set os.Stdout to /dev/null like this:
```
stdout = os.Stdout
os.Stdout, err = os.Open(os.DevNull)
```
so nothing can be written to stdout.

And then at the end of the command execution, we explicitly write json string to os.Stdout
```
out, err := json.Marshal(<-sc)
io.WriteString(stdout, string(out))
```

In this way, any output to os.Stdout during command execution is suppressed, so adding a new fmt.Println line in future won't break json output. 

Are you ok with this implementation?

**Comment by anson627** [comment]
I will add a string parameter called **--output** for output format (mimic kubectl).  And json format is set when the parameter has value **json**. Is that OK?

**Comment by anson627** [comment]
> I will add a string parameter called **--output** for output format (mimic kubectl). And json format is set when the parameter has value **json**. Is that OK?

ping @xiang90  I want to make sure you are ok with this parameter name before I update the PR

**Comment by anson627** [comment]
@xiang90 I updated the PR based on your feedback. Please help to take a look again. Thanks!

**Comment by anson627** [comment]
ping @xiang90

**Comment by anson627** [comment]
ping @gyuho can you help to take a look?

**Comment by anson627** [comment]
@spzala or @jingyih can you help to take a look?

File: tools/benchmark/cmd/encoder.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
+// Copyright 2015 The etcd Authors
```
Comment (spzala) [inline tools/benchmark/cmd/encoder.go:1]: nit, I would suggest to use 2020 here instead of 2015.

**Comment by anson627** [comment]
> @anson627 - naming `--output` sounds OK to me. Meanwhile we hear from @xiang90 couple of thoughts - do you think the flag can be a global flag (flags defined in the `tools/benchmark/cmd/root.go`) instead of enabling it individually? And secondly, considering there is no test cases, can you please share some outputs from your changes. Also had one inline comment. Thanks!

Sorry for the delayed replied! I was working on something else and didn't notice your comment until now.

It only make sense for `--output` to be a global flag, until every command support structured output. And I can move `--output` to root.go if that's preferred. I can also continue working on other commands to support structured output.

The output looks like this:
```
{
  "AvgTotal": 0.06699545499999998,
  "Fastest": 0.003297913,
  "Slowest": 0.020182931,
  "Average": 0.006699545499999998,
  "Stddev": 0.0217957758671141,
  "RPS": 39682539.68253969,
  "Total": 252,
  "ErrorDist": {},
  "Lats": [
    0.003297913,
    0.005040032,
    0.005160383,
    0.005290103,
    0.005538411,
    0.005568277,
    0.005600725,
    0.005657692,
    0.005658988,
    0.020182931
  ],
  "TimeSeries": null
}
```

File: tools/benchmark/cmd/encoder.go (hunk 1 above)
Comment (anson627) [reply tools/benchmark/cmd/encoder.go]: done

**Pull request closed without merging**